### ISSUE TITLE
Make servant-client more generic to support different ways of running a request

### ISSUE THREAD

**Issue by abailly** [haskell-servant/servant#798]
Currently, when one extracts a client from an API definition, she gets a tree of functions running inside `ClientM` monad:

```
type MyAPI = "foo" :> Capture "x" Int :> Get '[JSON] Foo
        :<|> "bar" :> Capture "y" String :> Get '[JSON] Bar

myAPI :: Proxy MyAPI
myAPI = Proxy

foo :: Int -> ClientM Foo
bar :: String -> ClientM Bar
foo :<|> bar = client myAPI
```

This implies that one needs to use  `runClientM foo env` to execute a query which uses http-client under the hood to actually perform the request in `IO`. However there are various use cases where one would like to change how the query is run:

1. Testing: one would like to provide canned answers and mock the server without having to fire up an actual HTTP server
2. Retry: In some circumstances, it would be desirable to retry failed requests according to some policy
3. Specialized run: One could want to tweak the requests in some ways or use different execution engines

A possible solution might be to parameterize the terminal combinators, e.g. `Verb`s that actually call `performRequests` and returns a `ClientM` action:

```
class MonadClient m where
  performRequestCT :: Proxy ct -> method -> Req -> m a
  performRequest     :: Proxy ct -> method -> Req -> m a

instance MonadClient ClientM where
 .... 

instance OVERLAPPING_
  -- Note [Non-Empty Content Types]
  ( MonadClient m, MimeUnrender ct a, BuildHeadersTo ls, ReflectMethod method, cts' ~ (ct ': cts)
  ) => HasClient (Verb method status cts' (Headers ls a)) where
  type Client (Verb method status cts' (Headers ls a))
    = m (Headers ls a)
  clientWithRoute Proxy req = do
    let method = reflectMethod (Proxy :: Proxy method)
    (hdrs, resp) <- performRequestCT (Proxy :: Proxy ct) method req
    return $ Headers { getResponse = resp
                     , getHeadersHList = buildHeadersTo hdrs
                     }
```

Another possible route would be to abstract the building of a request and provide different interpretations, something like what's done in https://github.com/soenkehahn/servant-request/

**Comment by abailly** [comment]
I have started working on a PR for this feature but got stuck while trying to add `m` parameter in `Client` type family: https://github.com/abailly/servant/blob/master/servant-client/src/Servant/Client.hs

**Comment by jkarni** [comment]
Great - this is something many people have wanted!

I'm not exactly sure what problem you're running into, but:

```
class HasClient api where
  type Client (m :: * -> *) (api :: *) :: *
  clientWithRoute :: Proxy api -> Req -> Client m api
```

This doesn't feel quite right - the `m` in `clientWithRoute` isn't constrained.

Does it help to have something like this?

```
class HasClient api where
  type Client (m :: * -> *) (api :: *) :: *
  clientWithRoute :: MonadClient m => Proxy api -> Req -> Client m api
```

Note that this approach doesn't allow for combinators to only make sense in
certain MonadClients. I don't think that's a problem for now, and I don't
think we should be worried about solving non-problems, but if does become one,
we could add another parameter to HasClient to settle the constraints.

We might also need a proxy for `m`, but I think not?

**Comment by abailly** [comment]
I have not created a `MonadClient` but a `RunClient`:

```
class (Monad m) => RunClient m ct result where
  runRequest :: MimeUnrender ct result
             => Proxy ct
             -> Method -> Req -> m result
```

which is more complex. I got a whole lot of:

```
    • Couldn't match expected type ‘Client m api’
                  with actual type ‘Client m0 api’
      NB: ‘Client’ is a type function, and may not be injective
      The type variable ‘m0’ is ambiguous
```
which makes sense because the `m` appears only in the result type of `clientWithRoute` as you noticed. I was just trying to get something quick and dirty to compile and see if I could extract a class easily but apparently this is not the case. I will try again later on this week.

Another approach that might be better has been suggested by @amarpotghan : Have `HasClient` generate a value in a `Request a` algebra capturing the request in a richer datatype than `Req` that can be interpreted by different backends.

**Comment by phadej** [comment]
This is a duplicate of https://github.com/haskell-servant/servant/issues/733

I'll comment more during this week. Great someone had time to pick this up!

**Comment by abailly** [comment]
It's funny that I have started writing code similar to existing proposals! Feel free to close this issue in favour of existing ones, of course.

**Comment by phadej** [comment]
@abailly could you open a PR, so it's easier to see the changes?

Few comments about `RunClient`:
- why MultiParamTypeClass? The lack of functional dependencies is causing the type-inference errors, as is often with MPTC.
- Remember `NoContent` (and `Headers`, and possible yet-to-be-implemented `StreamContent`)

**Comment by abailly** [comment]
This was really a quick and dirty attempt, and it does not even compile... Will try to make it compile then open a PR.

**Comment by jkarni** [comment]
Done in #803 (will be released as 0.12).

Thanks again!
